cub/device/dispatch/dispatch_batch_memcpy.cuh

```diff
@@ -0,0 +1,631 @@
+          typename BlockOffsetT,
+          typename BLevBufferOffsetTileState,
+          typename BLevBlockOffsetTileState>
+__launch_bounds__(int(ChainedPolicyT::ActivePolicy::AgentSmallBufferPolicyT::BLOCK_THREADS)) __global__
+  void BatchMemcpyKernel(InputBufferIt input_buffer_it,
+                         OutputBufferIt output_buffer_it,
+                         BufferSizeIteratorT buffer_sizes,
+                         BufferOffsetT num_buffers,
+                         BlevBufferSrcsOutItT blev_buffer_srcs,
+                         BlevBufferDstsOutItT blev_buffer_dsts,
+                         BlevBufferSizesOutItT blev_buffer_sizes,
+                         BlevBufferTileOffsetsOutItT blev_buffer_tile_offsets,
+                         BLevBufferOffsetTileState blev_buffer_scan_state,
+                         BLevBlockOffsetTileState blev_block_scan_state)
+{
+  static_assert(std::is_pointer<typename std::iterator_traits<InputBufferIt>::value_type>::value);
+  static_assert(std::is_pointer<typename std::iterator_traits<OutputBufferIt>::value_type>::value);
+
+  // Internal type used for storing a buffer's size
+  using BufferSizeT = typename std::iterator_traits<BufferSizeIteratorT>::value_type;
+
+  // Alias the correct tuning policy for the current compilation pass' architecture
+  using AgentBatchMemcpyPolicyT = typename ChainedPolicyT::ActivePolicy::AgentSmallBufferPolicyT;
+
+  // Block-level specialization
+  using AgentBatchMemcpyT = AgentBatchMemcpy<AgentBatchMemcpyPolicyT,
+                                             InputBufferIt,
+                                             OutputBufferIt,
+                                             BufferSizeIteratorT,
+                                             BufferOffsetT,
+                                             BlevBufferSrcsOutItT,
+                                             BlevBufferDstsOutItT,
+                                             BlevBufferSizesOutItT,
+                                             BlevBufferTileOffsetsOutItT,
+                                             BlockOffsetT,
+                                             BLevBufferOffsetTileState,
+                                             BLevBlockOffsetTileState>;
+
+  // Shared memory for AgentBatchMemcpy
+  __shared__ typename AgentBatchMemcpyT::TempStorage temp_storage;
+
+  // Process this block's tile of input&output buffer pairs
+  AgentBatchMemcpyT(temp_storage,
+                    input_buffer_it,
+                    output_buffer_it,
+                    buffer_sizes,
+                    num_buffers,
+                    blev_buffer_srcs,
+                    blev_buffer_dsts,
+                    blev_buffer_sizes,
+                    blev_buffer_tile_offsets,
+                    blev_buffer_scan_state,
+                    blev_block_scan_state)
+    .ConsumeTile(blockIdx.x);
+}
+
+template <typename BufferOffsetT, typename BlockOffsetT>
```

Comment:
Is there a reason to separate these types? Can we use a single `OffsetT` type instead?

Reply:
True! We don't need to specialise here. Fixed 👍